### ISSUE TITLE
Support -Z sanitizer=address

### ISSUE THREAD

**Issue by pquux** [mmastrac/rust-ctor#13]
Hey,

I'm not sure if this is a bug first, or if it's a bug in rust-ctor second, but rust ctor was how I encountered it so I'll file it here and youre welcome to close it or refile it, or direct me to refile it elsewhere.

So I was interested in using your crate to initialize static mutable globals. This is explicitly disallowed (perhaps document why?), but when searching around there is a helpful thread on github about depreciating static mut in favor of an UnsafeCell wrapper, so I used that to create a little library called `ctor_link`:

```rust
#[macro_use]
extern crate ctor;

use std::cell::UnsafeCell;
use std::collections::BTreeMap;

pub struct SyncUnsafeCell<T>(pub UnsafeCell<T>);
unsafe impl<T> Sync for SyncUnsafeCell<T> {}

#[ctor]
pub static X: SyncUnsafeCell<BTreeMap<u64, u64>> = {
    SyncUnsafeCell(UnsafeCell::new(BTreeMap::new()))
};
```

However, when I try to build this with address sanitizer:

```
$ RUSTFLAGS="-Z sanitizer=address" cargo build --target x86_64-unknown-linux-gnu
/* ... */
  = note: /usr/bin/ld: error: /home/x/ctor-link/target/x86_64-unknown-linux-gnu/debug/deps/ctor_link.3uc9gxsm5s9ojg2l.rcgu.o: size of section .ctors[_ZN9ctor_link20X___rust_ctor___ctor17ha03424299f229295E$93ae5017b04441301e73b201a904764c] is not multiple of address size
          /usr/bin/ld: final link failed: bad value
          collect2: error: ld returned 1 exit status
```

[This](https://github.com/espressif/binutils-esp32ulp/blob/master/bfd/elflink.c#L10264) code seems to indicate this error message has nothing to do with alignment, and instead is saying that the ctor section doesn't match the number of relocs. Some fiddling around with objdump and readelf verifies this:

```
$ objdump -h  /home/x/ctor-link/target/x86_64-unknown-linux-gnu/debug/deps/ctor_link.3uc9gxsm5s9ojg2l.rcgu.o |rg ctors
 14 .ctors        00000040  0000000000000000  0000000000000000  00000500  2**5
$ readelf -r  /home/x/ctor-link/target/x86_64-unknown-linux-gnu/debug/deps/ctor_link.3uc9gxsm5s9ojg2l.rcgu.o |rg ctors
Relocation section '.rela.ctors' at offset 0x23e8 contains 1 entry:
```

So the generated object has a single relocation, but is 0x40 bytes in size, leading to that error being thrown. Im guessing most of this can just be chalked up to different fringe language features colliding in an unexpected way and thus isn't really a serious problem, but thought I'd pass it on and let you make that determiniation.

Otherwise, thanks for the crate -- I was super happy when I found it as it's really useful for FFI situations.

**Comment by mmastrac** [comment]
Just to clarify- is this 0.1.7 or 0.1.8? I definitely had a correctness bug in 0.1.7 but it's possible it's still around.

**Comment by mmastrac** [comment]
Ahh - I think I understand what's going on. Looks like address sanitizer is padding the function object.

Making the static function extern _might_ help - I'll give this a shot when I have a second to put together a test case on macOS:

```
diff --git a/ctor/src/lib.rs b/ctor/src/lib.rs
index 284d8f2..adf995e 100644
--- a/ctor/src/lib.rs
+++ b/ctor/src/lib.rs
@@ -205,8 +205,8 @@ pub fn ctor(_attribute: TokenStream, function: TokenStream) -> TokenStream {
             #[cfg_attr(windows, link_section = ".CRT$XCU")]
             static #ctor_ident
             :
-            unsafe fn() = {
-                unsafe fn initer() {
+            unsafe extern fn() = {
+                unsafe extern fn initer() {
                     #storage_ident = Some(#expr);
                 }; initer }
             ;
```

**Comment by mmastrac** [comment]
I dug into this a bit and I think that this is going to be a rust core issue. Looks like address sanitizer increases the size of the symbol, which is incompatible with the layout of the `.ctors` section. To fix this I would probably need to add some sort of layer of indirection through a C shim that wasn't subject to address sanitization.

I don't fully understand this, but maybe @japaric could chime in.

**Comment by pquux** [comment]
Cool, thanks for looking into it.

I solved this basically by removing Asan -- I was only using it to sanitize my FFI code and it wasn't actually working there so easier just to ditch it entirely.

Out of curiosity, based on your patch shouldn't the init be extern anyways as it will be called by the loader, which will assuming a C abi?

**Comment by mmastrac** [comment]
> Out of curiosity, based on your patch shouldn't the init be extern anyways as it will be called by the loader, which will assuming a C abi?

I _think_ that's correct but there's some subtlety around this code that I need to understand better. It might be working purely because it's a function that takes no arguments and that's the same for the C ABI and Rust's internal conventions. :)

**Comment by mmastrac** [comment]
Marking this as `blocked` as it will need some upstream work. I _think_ we'll need to get upstream Rust to add an attribute to ensure that functions like this aren't padded. I'll keep considering alternative solutions but so far I'm coming up black.

**Comment by mmastrac** [comment]
Can you see if #19 fixes this?

**Comment by mmastrac** [comment]
Confirming fixed in #29
